### PR TITLE
feat(locking): integrated step based locking of properties

### PR DESCRIPTION
**Pull request by SimonTheSourcerer**
feat(locking): added manual unlock list to allows explicit unlocks
feat(locking): moved LockableProperty into Core
feat: add tabs to the Step Inspector
fix: NRE on scripts recompilation when you have the Step Inspector opened.
fix: performance issues with the creator window.
test: add unit tests for autolocking feature (#103)
ci: removed all files required for jenkins (#102)
refactor: reworked the runtime configuration

### Description

**Features**
* Reworked how the lock system works, added a new lock strategy which is based upon LockableProperty to lock SceneObject's properties according to usage in steps. TRNG-55

**Improvements**
* Improved performance in Chapter selection UI
* StepInspector UI now splits behaviors and conditions into different tabs.

**Chores**
* Removed Jenkins files from repository, as switching to Azure is complete.

### Type of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

### How Has This Been Tested?
* We added new unit tests
* We manual tested

**Test Configuration**:
Also update:
- Basic-Interaction
- Basic-Conditions & Behaviors
- VRTK-Interaction

### CODE REVIEW

File: Editor/UI/Drawers/TabsGroupDrawer.cs [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+    {
+        public override Rect Draw(Rect rect, object currentValue, Action<object> changeValueCallback, GUIContent label)
+        {
+            ITabsGroup tabsGroup = (ITabsGroup)currentValue;
+
+            // Draw tabs selector.
+            float tabsHeight = DrawToolbox(rect, tabsGroup, changeValueCallback).height;
+
+
+            // Get drawer for the object under the tab.
```
Comment (SimonTheSourcerer) [inline Editor/UI/Drawers/TabsGroupDrawer.cs:20]:
```suggestion
            // Get drawer for the object under the tab.
```

File: Runtime/Properties/ILockablePropertiesProvider.cs [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+
+namespace Innoactive.Creator.Core
+{
+    /// <summary>
+    /// This interface is used to allow entities, for example <see cref="Transition"/> or <see cref="Conditions"/>
+    /// to provide properties which should be locked.
+    /// </summary>
+    public interface ILockablePropertiesProvider
+    {
+        IEnumerable<LockablePropertyData> GetLockableProperties();
```
Comment (SimonTheSourcerer) [inline Runtime/Properties/ILockablePropertiesProvider.cs:12]: docu missing

File: Runtime/RestrictiveEnvironment/DefaultStepLockHandling.cs [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+            unlockList = unlockList.Union(manualUnlocked);
+
+            foreach (LockablePropertyData lockable in unlockList)
+            {
+                lockable.Property.SetLocked(false);
+            }
+        }
+
+        /// <inheritdoc />
+        public override void Lock(IStepData data, IEnumerable<LockablePropertyData> manualUnlocked)
```
Comment (SimonTheSourcerer) [inline Runtime/RestrictiveEnvironment/DefaultStepLockHandling.cs:25]: Some comments could improve the understanding what is happening here.

File: Runtime/Properties/PropertyReflectionHelper.cs [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+using Innoactive.Creator.Core.SceneObjects;
+using Innoactive.Creator.Core.Utils;
+using UnityEngine;
+
+namespace Innoactive.Creator.Core
+{
+    /// <summary>
+    /// Helper class which provides methods to extract <see cref="LockablePropertyData"/> from different course entities.
+    /// </summary>
+    public static class PropertyReflectionHelper
```
Comment (SimonTheSourcerer) [inline Runtime/Properties/PropertyReflectionHelper.cs:18]: Should be internal

File: Editor/UI/Drawers/TabsGroupDrawer.cs [hunk 5]
```diff
@@ -0,0 +1,65 @@
+using System;
+using System.Linq;
+using Innoactive.Creator.Core.Tabs;
+using UnityEditor;
+using UnityEngine;
+
+namespace Innoactive.CreatorEditor.UI.Drawers
+{
+    [DefaultTrainingDrawer(typeof(ITabsGroup))]
+    internal class TabsGroupDrawer : AbstractDrawer
```
Comment (Gusinuhe) [inline Editor/UI/Drawers/TabsGroupDrawer.cs:10]: Documentation would be nice

File: Editor/UI/Drawers/TabsGroupDrawer.cs [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+                    {
+                        tabsGroup.Selected = selected;
+                        return tabsGroup;
+                    },
+                    () =>
+                    {
+                        tabsGroup.Selected = oldSelected;
+                        return tabsGroup;
+                    },
+                    changeValueCallback);
```
Comment (Gusinuhe) [inline Editor/UI/Drawers/TabsGroupDrawer.cs:59]: Weird indentation

File: Editor/UI/Windows/StepWindow.cs [hunk 7]
```diff
@@ -78,6 +95,7 @@ private void ModifyStep(object newStep)
         {
             step = (IStep)newStep;
             GlobalEditorHandler.CurrentStepModified(step);
+            StepChanged?.Invoke(this, new StepModifiedEventArgs(step));
```
Comment (Gusinuhe) [inline Editor/UI/Windows/StepWindow.cs:98]: Is this necessary? Do we use it anywhere?

File: Runtime/Configuration/RuntimeConfigWrapper.cs [hunk 8]
```diff
@@ -0,0 +1,39 @@
+﻿using System;
+using Innoactive.Creator.Core.SceneObjects;
+using UnityEngine;
+
+namespace Innoactive.Creator.Core.Configuration
+{
+    /// <summary>
+    /// This is wrapper used for <see cref="IRuntimeConfiguration"/> configurations, which
```
Comment (Gusinuhe) [inline Runtime/Configuration/RuntimeConfigWrapper.cs:8]: Either

```suggestion
    /// This wrapper is used for <see cref="IRuntimeConfiguration"/> configurations, which
```
or 

```suggestion
    /// This is a wrapper used for <see cref="IRuntimeConfiguration"/> configurations, which
```

File: Runtime/Properties/LockableProperty.cs [hunk 9]
```diff
@@ -0,0 +1,96 @@
+﻿﻿using System;
+ using Innoactive.Creator.Core.SceneObjects;
+ using UnityEngine;
+
+namespace Innoactive.Creator.Core.Properties
+{
+    /// <summary>
+    /// TrainingSceneObjectProperty which is lockable, to allow the restrictive environment to handle
```
Comment (Gusinuhe) [inline Runtime/Properties/LockableProperty.cs:8]:
```suggestion
    /// <see cref="TrainingSceneObjectProperty"/> which is lockable, to allow the restrictive environment to handle
```

File: Runtime/Properties/LockableProperty.cs [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+    /// locking/unlocking your properties, extend this class.
+    /// </summary>
+    public abstract class LockableProperty : TrainingSceneObjectProperty, ILockable
+    {
+        public event EventHandler<LockStateChangedEventArgs> Locked;
+        public event EventHandler<LockStateChangedEventArgs> Unlocked;
```
Comment (Gusinuhe) [inline Runtime/Properties/LockableProperty.cs:14]: Missing docs

File: Runtime/Properties/LockableProperty.cs [hunk 11]
```diff
@@ -0,0 +1,96 @@
...
+
+        [SerializeField]
+        private bool lockOnParentObjectLock = true;
+
+        public bool LockOnParentObjectLock
```
Comment (Gusinuhe) [inline Runtime/Properties/LockableProperty.cs:19]: Missing docs

File: Runtime/Properties/LockablePropertyData.cs [hunk 12]
```diff
@@ -0,0 +1,47 @@
...
+    /// <summary>
+    /// Contains a target <see cref="LockableProperty"/> and additional information which define how the property is handled.
+    /// </summary>
+    public class LockablePropertyData
+    {
+
+        /// <summary>
+        /// Target lockable property.
+        /// </summary>
+        public readonly LockableProperty Property;
```
Comment (Gusinuhe) [inline Runtime/Properties/LockablePropertyData.cs:14]:
```suggestion
        /// <summary>
        /// Target lockable property.
        /// </summary>
        public readonly LockableProperty Property;
```

File: Runtime/Properties/LockablePropertyData.cs [hunk 13]
```diff
@@ -0,0 +1,47 @@
...
+            if (ReferenceEquals(null, obj)) return false;
+            if (ReferenceEquals(this, obj)) return true;
+            if (obj.GetType() != this.GetType()) return false;
+            return Equals((LockablePropertyData) obj);
+        }
+
+        public override int GetHashCode()
+        {
+            return (Property != null ? Property.GetHashCode() : 0);
+        }
```
Comment (Gusinuhe) [inline Runtime/Properties/LockablePropertyData.cs:45]: Missing docs

File: Runtime/Properties/LockablePropertyReference.cs [hunk 14]
```diff
@@ -0,0 +1,60 @@
...
+    /// Serializable reference to a <see cref="LockableProperty"/>
+    /// </summary>
+    [DataContract(IsReference = true)]
+    public class LockablePropertyReference
+    {
+        [DataMember]
+        public SceneObjectReference Target;
+
+        [DataMember]
+        public string Type;
```
Comment (Gusinuhe) [inline Runtime/Properties/LockablePropertyReference.cs:18]: Missing docs

File: Runtime/Properties/PropertyReflectionHelper.cs [hunk 15]
```diff
@@ -0,0 +1,110 @@
...
+    {
+        /// <summary>
+        /// Extracts a list of <see cref="LockablePropertyData"/> from given Step.
+        /// </summary>
+        public static List<LockablePropertyData> ExtractLockablesFromStep(IStep step)
```
Comment (Gusinuhe) [inline Runtime/Properties/PropertyReflectionHelper.cs:23]: Is this method used? 🤔

File: Runtime/RestrictiveEnvironment/DefaultStepLockHandling.cs [hunk 16]
```diff
@@ -0,0 +1,77 @@
...
+namespace Innoactive.Creator.Core.RestrictiveEnvironment
+{
+    /// <summary>
+    /// Restricts interaction with scene objects by using LockableProperties, which are extracted from the <see cref="IStepData"/>.
+    /// </summary>
+    public class DefaultStepLockHandling : StepLockHandlingStrategy
+    {
+
+        /// <inheritdoc />
+        public override void Unlock(IStepData data, IEnumerable<LockablePropertyData> manualUnlocked)
```
Comment (Gusinuhe) [inline Runtime/RestrictiveEnvironment/DefaultStepLockHandling.cs:13]:
```suggestion
        /// <inheritdoc />
        public override void Unlock(IStepData data, IEnumerable<LockablePropertyData> manualUnlocked)
```

File: Runtime/Tabs/DynamicTab.cs [hunk 17]
```diff
@@ -0,0 +1,36 @@
...
+
+        public object GetValue()
+        {
+            return getter();
+        }
+
+        public void SetValue(object value)
+        {
+            setter(value);
+        }
```
Comment (Gusinuhe) [inline Runtime/Tabs/DynamicTab.cs:21]: Missing docs

File: Runtime/Transition.cs [hunk 18]
```diff
@@ -139,5 +141,18 @@ public Transition()
                 };
             }
         }
+
+        public IEnumerable<LockablePropertyData> GetLockableProperties()
```
Comment (Gusinuhe) [inline Runtime/Transition.cs:145]: Missing docs

File: Runtime/Tabs/DynamicTab.cs (hunk 17 above)
Comment (SimonTheSourcerer) [reply Runtime/Tabs/DynamicTab.cs:21]: its internal

File: Editor/UI/Drawers/TabsGroupDrawer.cs (hunk 5 above)
Comment (SimonTheSourcerer) [reply Editor/UI/Drawers/TabsGroupDrawer.cs:10]: internal class which is a drawer, would except people to understand without docu :P

File: Runtime/Properties/LockableProperty.cs (hunk 9 above)
Comment (SimonTheSourcerer) [reply Runtime/Properties/LockableProperty.cs:8]: done

File: Runtime/Properties/PropertyReflectionHelper.cs (hunk 15 above)
Comment (SimonTheSourcerer) [reply Runtime/Properties/PropertyReflectionHelper.cs:23]: nopes

File: Runtime/Transition.cs (hunk 18 above)
Comment (SimonTheSourcerer) [reply Runtime/Transition.cs:145]: added

File: Runtime/Tabs/DynamicTab.cs (hunk 17 above)
Comment (Gusinuhe) [reply Runtime/Tabs/DynamicTab.cs:21]: I agree that is optional but it would be nice.

File: Runtime/BehaviorCollection.cs [hunk 19]
```diff
@@ -25,7 +25,7 @@ public class EntityData : EntityCollectionData<IBehavior>, IBehaviorCollectionDa
             /// List of all <see cref="IBehavior"/>s added.
             /// </summary>
             [DataMember]
-            [DisplayName("Behaviors"), Separated, Foldable, ListOf(typeof(FoldableAttribute), typeof(DeletableAttribute), typeof(SeparatedAttribute), typeof(DrawIsBlockingToggleAttribute)), ExtendableList]
+            [DisplayName(""), ListOf(typeof(FoldableAttribute), typeof(DeletableAttribute), typeof(DrawIsBlockingToggleAttribute)), ExtendableList]
```
Comment (tomwim) [inline Runtime/BehaviorCollection.cs:28]: Why no display name anymore? Asking for a friend

Comment (SimonTheSourcerer) [reply Runtime/BehaviorCollection.cs:28]: because the name is now displayed inside the tab instead :)

Comment (aleksei-korolev) [reply Runtime/BehaviorCollection.cs:28]: It duplicates the tab name.

**Pull request merged**